crates/pallets/starknet/src/lib.rs

```diff
@@ -1122,4 +1123,52 @@ impl<T: Config> Pallet<T> {
         };
         Ok(events)
     }
+
+    /// Compute the global state root and store it in the runtime storage.
+    /// This function is called at the end of each block.
+    /// It iterates through all the pending storage changes and updates the storage trie.
+    /// It then computes the state root and stores it in the runtime storage.
+    ///
+    /// # Returns
+    ///
+    /// The global state root.
+    pub fn compute_and_store_state_root() -> Felt252Wrapper {
+        // Update contracts trie
+        let mut commitments = State::<T>::get();
+        let pending_state = PendingState::<T>::iter();
+
+        pending_state.for_each(|(contract_address, storage_diffs)| {
+            // Retrieve state trie for this contract.
+            let mut state_tree = StorageTries::<T>::get(contract_address).unwrap_or_default();
+            // For each smart contract, iterate through storage diffs and update the state trie.
+            storage_diffs.into_iter().for_each(|(storage_key, storage_value)| {
+                state_tree.set(storage_key, storage_value);
+            });
+
+            // Update the state trie for this contract in runtime storage.
+            StorageTries::<T>::set(contract_address, Some(state_tree.clone()));
+
+            // We then compute the state root
+            // And update the storage trie
+            let state_root = state_tree.commit();
+
+            let nonce = Nonces::<T>::get(contract_address);
+            let class_hash = ContractClassHashes::<T>::get(contract_address).unwrap_or_default();
+            let hash = calculate_contract_state_hash::<T::SystemHash>(class_hash, state_root, nonce);
+            commitments.storage_commitment.set(contract_address, hash);
+
+            // Finally update the contracts trie in runtime storage.
+            State::<T>::mutate(|state| {
+                state.storage_commitment = commitments.clone().storage_commitment;
+            });
+
+            PendingState::<T>::remove(contract_address);
```

Comment:
Same. Once you finished iterating over all of the pending states, you can just kill the storage itself, just bellow your loop. Rather than removing each entry 1 by 1

Reply:
we can't kill the storage cause its a `StorageMap` :(

Reply:
Use this ? `PendingState::<T>::remove_all(None);`

Reply:
Does order matter to you?
Otherwise you can use drain and directly iterate on it.

`remove_all` is deprecated, you should use [`clear`](https://paritytech.github.io/substrate/master/frame_support/storage/types/struct.StorageMap.html#method.clear) instead, but you have to check the result and eventually call it again depending of the limit you used